File: 134-gas-station/134-gas-station.scala
object Solution {
    def canCompleteCircuit(gas: Array[Int], cost: Array[Int]): Int = {
        var total = 0
        var cur = 0
        val n = gas.length
        var start = 0

        for(i <- 0 to n - 1) {
            val gain = gas(i) - cost(i)
            total += gain
            cur += gain
            if (cur < 0) { 
                start = (i + 1) % n
                cur = 0
            }
        }

        if (total < 0) -1 else start
    }
}

File: 134-gas-station/134-gas-station.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int res = 0;
        int total = 0;
        
        for(int i = 0, cur = 0; i < gas.length; i++) {
            int g = gas[i];
            int c = cost[i];
            int gain = g - c;
            cur += gain;
            total += gain;
            if (cur < 0) {
                cur = 0;
                res = i + 1;
            }
        }
        
        return (total >= 0)? res : -1;
    }
}